### PR TITLE
Option to disable question order randomization on exams

### PR DESCRIPTION
**Pull request by ffund**
For #5278: Disable question order randomization on `Exam`-type assessments by settings `"shuffleQuestions": false` in the assessment config.
 
Default behavior is the same as before (if `shuffleQuestions` is not specified in the assessment config, questions are randomized on exams and not randomized on homeworks).

On exams, the question numbers are still "Question 1", "Question 2", etc. regardless of the `shuffleQuestions` setting, so it is not as obvious to students whether questions or randomized or not. (This is different from homeworks, where if `shuffleQuestions` is true, numbers are assigned to questions so that every student has the same question number for a given question, regardless of the order in which they see the question.)

### CODE REVIEW

File: sprocs/question_order.sql [hunk 1]
```diff
@@ -70,7 +70,11 @@ WINDOW
                         WHEN a.shuffle_questions THEN iq.order_by
                         ELSE aq.number
                     END
-                WHEN a.type = 'Exam' THEN iq.order_by
+                WHEN a.type = 'Exam' THEN
+                    CASE
+                        WHEN a.shuffle_questions THEN iq.order_by
+                        ELSE aq.number
+                    END
```
Comment (nwalters512) [inline sprocs/question_order.sql:77]: Let's collapse this into the previous case, e.g.

```sql
WHEN (a.type = 'Homework' OR a.type = 'Exam') THEN ...
```

In practice, we should only ever have this case, but it looks like there are some other (probably unsupported) assessment types from the deep past that I don't want to risk breaking: https://github.com/PrairieLearn/PrairieLearn/blame/15a01e6b15c520f5e0435ced1a1fb0d5bfb8aefd/database/enums/enum_assessment_type.pg

File: sync/fromDisk/assessments.js [hunk 2]
```diff
@@ -55,7 +55,11 @@ function getParamsForAssessment(assessmentInfoFile, questionIds) {
     number: assessment.number,
     title: assessment.title,
     multiple_instance: assessment.multipleInstance ? true : false,
-    shuffle_questions: assessment.shuffleQuestions ? true : false,
+    shuffle_questions:
```
Comment (nwalters512) [inline sync/fromDisk/assessments.js:58]: I'd encourage you to add a few tests for this behavior to https://github.com/PrairieLearn/PrairieLearn/blob/master/tests/sync/assessmentsSync.test.js:

* Check that `shuffleQuestions` defaults to `true` for Exam-type assessments
* Check that `shuffleQuestions` can be overridden to `false` for Exam-type assessments

**Review by nwalters512 — COMMENTED** [review]
Thanks for the contribution! Please add text like "fixes #5278" to the PR description so that that issues is automatically closed once this is merged. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword for docs on that feature.

File: sprocs/question_order.sql (hunk 1 above)
Comment (ffund) [reply sprocs/question_order.sql]: Done in [#40e6ff5](https://github.com/PrairieLearn/PrairieLearn/pull/5892/commits/40e6ff57456d3d2e36903d9b848ca27c6be91c72).

**Comment by ffund** [comment]
@nwalters512 I don't think #5278 should be closed yet because it says:

> We should enable this for Exam assessments on a per-zone and whole-assessment basis

and this PR only does it on a whole-assessment basis.

File: sync/fromDisk/assessments.js (hunk 2 above)
Comment (ffund) [reply sync/fromDisk/assessments.js:58]: Added the suggested tests in cede76e

File: tests/sync/assessmentsSync.test.js [hunk 3]
```diff
@@ -111,6 +111,50 @@ describe('Assessment syncing', () => {
     assert.equal(syncedData.assessment_questions[1].question.qid, util.ALTERNATIVE_QUESTION_ID);
   });
 
+  it('checks that shuffleQuestions defaults to true for exam-type assessments', async () => {
```
Comment (nwalters512) [inline tests/sync/assessmentsSync.test.js:114]: Nit for consistency with other test names:
```suggestion
  it('defaults shuffleQuestions to true for an Exam-type assessment', async () => {
```

File: tests/sync/assessmentsSync.test.js [hunk 4]
```diff
@@ -111,6 +111,50 @@ describe('Assessment syncing', () => {
...
+    const courseData = util.getCourseData();
+    const assessment = makeAssessment(courseData, 'Exam');
+    courseData.courseInstances[util.COURSE_INSTANCE_ID].assessments['newexam'] = assessment;
+    await util.writeAndSyncCourseData(courseData);
+
+    const syncedData = await getSyncedAssessmentData('newexam');
+    assert.isTrue(syncedData.assessment.shuffle_questions);
+  });
+
+  it('checks that shuffleQuestions can be set to false for exam-type assessments', async () => {
```
Comment (nwalters512) [inline tests/sync/assessmentsSync.test.js:124]:
```suggestion
  it('allows shuffleQuestions to be set to false for an Exam-type assessment', async () => {
```

File: tests/sync/assessmentsSync.test.js [hunk 5]
```diff
@@ -111,6 +111,50 @@ describe('Assessment syncing', () => {
...
+    assessment.shuffleQuestions = false;
+
+    courseData.courseInstances[util.COURSE_INSTANCE_ID].assessments['newexam'] = assessment;
+    await util.writeAndSyncCourseData(courseData);
+
+    const syncedData = await getSyncedAssessmentData('newexam');
+    assert.isFalse(syncedData.assessment.shuffle_questions);
+  });
+
+  it('checks that shuffleQuestions defaults to false for homework-type assessments', async () => {
```
Comment (nwalters512) [inline tests/sync/assessmentsSync.test.js:136]:
```suggestion
  it('defaults shuffleQuestions to false for a Homework-type assessment', async () => {
```

File: tests/sync/assessmentsSync.test.js [hunk 6]
```diff
@@ -111,6 +111,50 @@ describe('Assessment syncing', () => {
...
+    const courseData = util.getCourseData();
+    const assessment = makeAssessment(courseData, 'Homework');
+    courseData.courseInstances[util.COURSE_INSTANCE_ID].assessments['newhomework'] = assessment;
+    await util.writeAndSyncCourseData(courseData);
+
+    const syncedData = await getSyncedAssessmentData('newhomework');
+    assert.isFalse(syncedData.assessment.shuffle_questions);
+  });
+
+  it('checks that shuffleQuestions can be set to true for homework-type assessments', async () => {
```
Comment (nwalters512) [inline tests/sync/assessmentsSync.test.js:146]:
```suggestion
  it('allows shuffleQuestions to be set to true for a Homework-type assessment', async () => {
```

**Review by nwalters512 — APPROVED** [review]
Thanks for pointing out that #5278 also mentions supporting this on a per-zone basis too! This looks good; let me know when you're ready for me to merge.

**Comment by ffund** [comment]
@nwalters512 go ahead and merge, thank you!

**Comment by ffund** [comment]
@nwalters512 bumping in case you missed the notification last week: please go ahead and merge whenever you are ready.

**Comment by trombonekenny** [comment]
@ffund Asked about when this would be merged and deployed in OH today.

**Comment by nwalters512** [comment]
Sorry for the delay - landing this now!

**Pull request merged**